### DISCUSSION TITLE
Can the Pico W do a WiFi access point?

### DISCUSSION THREAD

**Discussion by fapplin** [earlephilhower/arduino-pico#701] [Q&A] [answered]
I was able to get connected to my local network with your WiFiServer example, but can I make the server an access point server? If so, how?

Thanks,

Frank

**Comment by earlephilhower** [comment]
Yes, AP is supported but I didn't get around to writing an example for it.  See https://github.com/earlephilhower/arduino-pico/blob/a97dc38e8778f531ec4d7eddfe66f464d8be434a/libraries/WiFi/src/WiFiClass.h#L79-L82 .

If you come up with a good simple example, feel free to do a PR to add it!

**Reply by fapplin** [reply]
OK. I tried making some mods to your code and I get connected to my cell phone, but it never seems to get beyond the WiFi.status() != WL_CONNECTED) code. Can you see what I'm doing wrong? I've been look at code from a NodeMCU github and the WiFi stuff is handled a bit different. see: [](https://github.com/binaryupdates/NodeMCU-Webserver-Access-Point/blob/main/ESP8266_Webserver_Access_Point_Mode.ino)
https://github.com/binaryupdates/NodeMCU-Webserver-Access-Point/blob/main/ESP8266_Webserver_Access_Point_Mode.ino
```
// Placed in the public domain by Earle F. Philhower, III, 2022

#include <WiFi.h>

#ifndef STASSID
#define STASSID "PicoW"
#define STAPSK "123456789"
#endif

const char* ssid = STASSID;
const char* password = STAPSK;

int port = 4242;
int dotCount;

WiFiServer server(port);

void setup() {
  Serial.begin(115200);
  //WiFi.mode(WIFI_AP);
  WiFi.beginAP(ssid, password);
  //WiFi.setHostname("PicoW2");
  //Serial.printf("Connecting to '%s' with '%s'\n", ssid, password);
  //WiFi.begin();
  dotCount = 0;
  while (WiFi.status() != WL_CONNECTED) {
    if (dotCount < 80) 
    {
      dotCount++;
      Serial.print(".");
    } 
    else 
    {
      Serial.println("");
      dotCount = 0;
      delay(100);
    } //end if
    
  }
  Serial.printf("\nConnected to WiFi\n\nConnect to server at %s:%d\n", WiFi.localIP().toString().c_str(), port);

  server.begin();
}

void loop() {
  static int i;
  delay(1000);
  Serial.printf("--loop %d\n", ++i);
  delay(10);
  WiFiClient client = server.available();
  if (!client) {
    return;
  }
  client.println("Type anything and hit return");
  while (!client.available()) {
    delay(10);
  }
  String req = client.readStringUntil('\n');
  Serial.println(req);
  client.printf("Hello from Pico-W\r\n");
  client.flush();
}
```

**Reply by earlephilhower** [reply]
There is no need to check for connected in AP mode...it's not connecting to anything, after all, it is becoming a hotspot.

I'll check and see what other cores do in the AP case.  Maybe they just return WL)CONNECTED anyway.

**Reply by fapplin** [reply]
OK. Success!! I strip it to the bare basics. Got the ip address, went to it with my phone's browser and got a response!!

```
// Placed in the public domain by Earle F. Philhower, III, 2022

#include <WiFi.h>

#ifndef STASSID
#define STASSID "PicoW"
#define STAPSK "123456789"
#endif

const char* ssid = STASSID;
const char* password = STAPSK;

int port = 80;
int dotCount;

WiFiServer server(port);

void setup() {
  Serial.begin(115200);

  WiFi.beginAP(ssid, password);
 
  Serial.println(WiFi.localIP());

  server.begin();
}

void loop() {
  static int i;
  delay(1000);
  Serial.printf("--loop %d\n", ++i);
  Serial.println(WiFi.localIP());
  delay(10);
  WiFiClient client = server.available();

  String req = client.readStringUntil('\n');
  Serial.println(req);
  client.printf("Hello from Pico-W\r\n");
  client.flush();
}
```

**Reply by fapplin** [reply]
O, Phil, I hope you can help out again. It seems like the many arduino wifi libraries handle things different. I would like to be able to handle a form http POST. This was one example that I was looking at for an ESP8266, but when I look at your code - you don't have a server.on() method for handling callbacks. 
[](url)https://circuits4you.com/2019/03/20/esp8266-receive-post-get-request-data-from-website/

So, how can I interrogate an http post? I see the POST when I do:

```
 String req = client.readStringUntil('\n');
  Serial.println(req);
```
but, how do I interrogate beyond that to the fields on my form?

Thanks again,

Frank

**Reply by earlephilhower** [reply]
Simplest way, you need an HTTP server library, really, to handle that.  I have not (yet) included one, but any that work with standard Arduino should be able to handle it.

You could also manually parse the `POST` but that is actually more complicated than you'd think, due to the way modern browsers send lots of info/etc. unrelated to HTTP/1.0 POST.

**Reply by fapplin** [reply]
Yeah, I've seen a good parser that I will use, but how do I get the data from my form? What function/method call? 

String req = client.readStringUntil('\n');

only returns the string POST /mypage HTTP/1.1

**Reply by earlephilhower** [reply]
The parser should handle it.  It will do `client.read()` not `readUntil` (which in your case only reads until a newline, whereas the actual POST data is encoded afterwards on subsequent lines).

**Reply by fapplin** [reply]
And this is the parser I was looking at using.

[https://github.com/NatanBiesmans/Arduino-POST-HTTP-Parser](url)

**Comment by fapplin** [comment, marked as answer]
OK. Got it working by modifying the library mentioned above and including the files in my sketch!

Thanks.

**Comment by fapplin** [comment]
Are you experiencing problems obtaining an IP address using access point? I watch it on my phone trying to obtain IP address and it's like it gets stuck and then gives up. Why do you need an IP address if it seems to be always 192.168.42.1? Can we set our own address?

I was making good progress, today, but it seems my Pico just keeps disconnecting to reconnect to obtain an IP address.  I've tried it on another Pico W - same results.

**Comment by earlephilhower** [comment]
I think there might be some confusion here.  In AP mode, the Pico is not connected to the internet and doesn't have any address other than 192.168.42.1 (well, that's configurable but it's fixed).  So the Pico's not going to try and connect to the internet.  Your phone, however, might notice "Hey, this AP (the PicoW) I connected to doesn't have inet connection...let me go back to 4G"

**Reply by earlephilhower** [reply]
Also, you don't need to be in AP mode to use a WiFiServer.  Just start one in normal STA and use the Pico's given IP as the address...

**Reply by fapplin** [reply]
I know it's not trying to connect to the Internet. Maybe, I need to stop looking at examples that are connected to Ethernet or WiFi (WIFI_STA) because I think they are just confusing me.

I basically want to connect to the access point on my phone, bring up a browser on my phone, go to 192.168.42.1, have a page with a form displayed, make some choices via radio buttons, press submit, and parse the values. I had it working, but I just think I'm doing more than needs doing and that's what is messing me up.

**Reply by fapplin** [reply]
OK, Phil, maybe I don't understand access point. I can do the following code (which is basically from Martyn Currey) connected to my wifi router. I can access the page (at the IP address) via my computer or my phone (connected to the router via WIFI). 

If I change WIFI_STA to WIFI_AP (and id and pwd), compile it, add it my myPico W, connect to it via my phone's WiFi, shouldn't I be able to access the page (at the IP address 192.168.42.1) - just like I did for WIFI_STA mode? Or, is this what I'm missing? I'm sorry for taking your time. I am a long time programmer - just fairly new to this stuff.

```
#include <WiFi.h>

const char WiFiPassword[] = "MyRouterPwd";
const char AP_NameChar[] = "MyRouter" ;
 
WiFiServer server(80);
 
String header = "HTTP/1.1 200 OK\r\nContent-Type: text/html\r\n\r\n";
String html_1 = "<!DOCTYPE html><html><head><title>LED Control</title></head><body><div id='main'><h2>LED Control</h2>";
String html_2 = "<form id='F1' action='LEDON'><input class='button' type='submit' value='LED ON' ></form><br>";
String html_3 = "<form id='F2' action='LEDOFF'><input class='button' type='submit' value='LED OFF' ></form><br>";
String html_4 = "</div></body></html>";
 
String request = "";
 
void setup() 
{
    //pinMode(LED_Pin, OUTPUT); 
   Serial.begin(115200);
 
  WiFi.mode(WIFI_STA);
  WiFi.setHostname("PicoW2");
  WiFi.begin(AP_NameChar, WiFiPassword);
  server.begin();
  delay(10000);
 
} // void setup()
 
 
void loop() 
{
 
    // Check if a client has connected
    WiFiClient client = server.available();

    //if (!client)  {  return;  } //this doesn't work at all

    Serial.printf("\nConnected to WiFi\n\nConnect to server at %s\n", WiFi.localIP().toString().c_str());

    // Read the first line of the request
    request = client.readStringUntil('\r');
 
    if (request.indexOf("LEDON") > 0)
    { 
      Serial.println("On");  
    }
    else if ( request.indexOf("LEDOFF") > 0 ) 
    { 
      Serial.println("Off");   
    } //end if
 
    client.flush();
 
    client.print( header );
    client.print( html_1 );
    client.print( html_2 );
    client.print( html_3 );
    client.print( html_4);
 
    delay(5);
  // The client will actually be disconnected when the function returns and 'client' object is detroyed
 
} // void loop()
```
